rust/ql/lib/codeql/rust/elements/internal/FormatConstructor.qll

```diff
@@ -0,0 +1,78 @@
+/**
+ *  This module defines the hook used internally to tweak the characteristic predicate of
+ *  `Format` synthesized instances.
+ *  INTERNAL: Do not use.
+ */
+
+private import codeql.rust.elements.internal.generated.Raw
+
+/**
+ *  The characteristic predicate of `Format` synthesized instances.
+ *  INTERNAL: Do not use.
+ */
+predicate constructFormat(Raw::FormatArgsExpr parent, int index) {
+  formatElement(parent, index, _).regexpMatch(formatRegex())
+}
+
+/**
+ * Match an element of a format string, either text (`Hello`) or a format placeholder (`{}`).
+ */
+string formatElement(Raw::FormatArgsExpr parent, int occurrenceIndex, int occurrenceOffset) {
+  result =
+    parent
+        .getTemplate()
+        .(Raw::LiteralExpr)
+        .getTextValue()
+        // TODO: should also handle surrounding quotes and escaped characters
+        .regexpFind(textRegex() + "|" + formatRegex(), occurrenceIndex, occurrenceOffset)
```

Comment:
To avoid mis-parsing something like `{{}}` . This is a `textRegex` even though the inner `{}` could match `formatRegex`.

Reply:
Ah, it's because `regexpFind` basically encodes the
```
format_string := text [ maybe_format text ] *
```
bit of the grammar. (I was confused why it was not something like `"(" + textRegex() + "|" + formatRegex() + ")+"`

Reply:
Indeed, I had something like `(text|format)*` at first, but capture groups inside a repetition only capture the last occurrence, which is not great. That's why I'm using `regexpFind` instead to find all occurences.